octue/cloud/pub_sub/service.py

```diff
@@ -187,23 +188,43 @@ def ask(self, service_id, input_values, input_manifest=None):
     def wait_for_answer(self, subscription, timeout=30):
         """Wait for an answer to a question on the given subscription, deleting the subscription and its topic once
         the answer is received.
+
+        :param octue.cloud.pub_sub.subscription.Subscription subscription: the subscription for the question's answer
+        :param float timeout: how long to wait for an answer before raising a TimeoutError
+        :raise TimeoutError: if the timeout is exceeded
+        :return dict: dictionary containing the keys "output_values" and "output_manifest"
         """
-        pull_response = self.subscriber.pull(
-            request={"subscription": subscription.path, "max_messages": 1},
-            timeout=timeout,
-            retry=create_custom_retry(timeout),
-        )
+        start_time = time.perf_counter()
+        no_message = True
+        attempt = 1
+
+        with self.subscriber:
+            while no_message:
+                logger.debug("Pulling messages from Google Pub/Sub: attempt %d.", attempt)
+
+                pull_response = self.subscriber.pull(
+                    request={"subscription": subscription.path, "max_messages": 1},
+                    retry=create_custom_retry(timeout),
+                )
+
+                try:
+                    answer = pull_response.received_messages[0]
+                    no_message = False
+
+                except IndexError:
+                    logger.debug("Google Pub/Sub pull response timed out early.")
+                    attempt += 1
 
-        try:
-            answer = pull_response.received_messages[0]
-        except IndexError:
-            raise TimeoutError("No answer received from topic %r", subscription.topic.path)
+                    if (time.perf_counter() - start_time) > timeout:
+                        raise TimeoutError(
+                            f"No answer received from topic {subscription.topic.path!r} after {timeout} seconds.",
+                        )
 
-        self.subscriber.acknowledge(request={"subscription": subscription.path, "ack_ids": [answer.ack_id]})
-        logger.debug("%r received a response to question on topic %r", self, subscription.topic.path)
+            self.subscriber.acknowledge(request={"subscription": subscription.path, "ack_ids": [answer.ack_id]})
+            logger.debug("%r received a response to question on topic %r.", self, subscription.topic.path)
```

Comment:
Shouldn't we be acking the pull response irrespective of whether it has a body? We'll end up with a lot of redeliveries of the empty messages...

Reply:
I'm not sure what you mean - is it that assuming the `answer` has an `ack_id` attribute is assuming it has a body?

Reply:
no, it's because the ack only happens after the try/catch so if an error is raised the message doesn't get acked.

Reply:
I don't think you can ack it unless there are messages in the pull response.

Reply:
i.e. you can only ack a message if you've successfully received it

Reply:
But that doesn't marry with their pattern of getting a pull response with no message body. What happens if you reproduce the error, but try to ack the pull response before it occurs? Does the ack complete or is there a different error?

Reply:
A pull response with no body isn't an empty message though, it's an absence of messages. Maybe I'm missing something? You can only acknowledge a message if you have its ID, which you only have if you receive the message though. If you don't have the ID, you can't pass it to
```
self.subscriber.acknowledge(
    request={"subscription": subscription.path, "ack_ids": [<ack ID>]}
)
```

Reply:
Ok let’s resolve this as it sounds like we’re talking at cross purposes re the meaning of ‘message’ and ‘body’. I assumed that what we were getting was a body, but with an empty message. Sounds like the ‘message’ I’m talking about is the ‘body you’re talking about, in which case your comments make sense here.